File: Least Prime Factor.cpp
class Solution {
  public:
    vector<int> leastPrimeFactor(int n) {
        // initialize the vector
        vector<int> ans(n+1);
        ans[1] = 1;
        
        // initialize all numbers as prime 
        for(int i = 2;i <= n; i++){
            ans[i]=i;
        }
        
        // sieve of eratosthenes
        for(int i = 2; i*i <= n; i++){
            // if prime, mark all multiples of i with their least prime factor
            if(ans[i]==i){
                for(int j = i*i; j <= n; j += i){
                    if(ans[j] == j){
                        ans[j] = i;
                    }
                }
            }
        }
        return ans;
    }
};

File: Least Prime Factor.java
class Solution{
    static boolean isPrime(int n){
        for(int i=2;i<=Math.sqrt(n);i++){
            if(n%i==0)return false;
        }
        return true;
    }
    public int[] leastPrimeFactor(int n){
        int ret[]=new int[n+1];
        ret[0]=0;
        ret[1]=1;
        for(int i=2;i<=n;i++){
            for(int j=2;j<=i;j++){
                if(i%j==0 && isPrime(j)){
                    ret[i]=j;
                    break;
                }
            }
        }
        return ret;
    }
}
